### ISSUE TITLE
Adding support for DX film edge barcode

### ISSUE THREAD

**Issue by Merinorus** [zxing-cpp/zxing-cpp#662]
# Adding support for DX film edge barcode
Hello ZXing team,

I'm working on a personal project where I would like to be able to scan DX film edge barcodes.

## DX film edge barcodes?

These are the kinds of barcodes we find on the edge of 35mm negative films, like at the bottom of the following image:
![Pasted image 20231114195712](https://github.com/zxing-cpp/zxing-cpp/assets/3023499/fda7de46-7c0b-4c63-b7e4-2767290fb4bd)

They refer to the [DX encoding system](https://en.wikipedia.org/wiki/DX_encoding).

In this barcode, there are two numbers:
- the film model (manufacturer and type of film)
- the current image number (+ a flag to indicate whether we are in the middle or the end of the image)

## So... Why would I need to read them?

More and more people are involved in digitizing negatives at home. The film model number is important to know, so we can apply the right color treatment to the pictures.

Some people use dedicated film scanners, which already read this code and can apply the appropriate color treatment automatically.

Other people, including me, use a DSLR to digitize the films. This is done in two main steps:
- Take a picture of the negative with a proper setup (backlight, macro lens...): with a lot of setups, only the picture itself is digitized, not the barcode.
- Apply an appropriate color treatment on the computer to retrieve more or less the original colors, to obtain the "positive". Darktable, Lightroom, Negative Lab Pro, etc. are some tools that can help in this way. Some tools can help in guessing the color treatment but, generally, it's better to know the film model.

This is why I'm interested in adding the support for DX film edge barcode. Some industrial products with closed-source software already do this kind of job to apply a color treatment, but the number itself is not displayed and I didn't find any open-source solution to do it.

## How to read them

DX edge barcodes have some characteristics in common with QR codes:
- patterns to indicate the beginning and the end of the barcode,
- a "clock" to help read the signal with the machine,
- the signal itself (2 dimensions in the QR codes, 1 dimension in the DX film edge barcode).

An expired [Kodak patent](https://patents.google.com/patent/US4965628A/en) explains how to read them ([pdf version](https://patentimages.storage.googleapis.com/d6/13/7d/51ea3bebaa8c99/US4965628.pdf)).

Here is an image extracted from the patent that explains the pattern. It does not explain exactly how to decode it but it gives an idea. If you are interested, I can post how to decode it step by step.

![Pasted image 20231108205211](https://github.com/zxing-cpp/zxing-cpp/assets/3023499/3334591a-5f68-4a2a-bb3a-8d0c5b214074)

## Does this belong to Zxing-CPP?

So my actual question is: does decoding this kind of barcode belong to this project? I understand this is a "niche" feature and a few people might use it. But, if I worked on this one and I opened a PR, would you consider it? Or do you know another project that would be more appropriate?

Thank you for your attention and sorry for the long post. At least I hope it was instructive!

**Comment by axxel** [comment]
Interesting subject. :). I had a look at your post and a few related Wikipedia pages and it seems to be a rather simple symbology that is expected to show up only in a very specific context. Am I right that any reasonable setup for digitization would result in images that basically look like your sample? Meaning: exactly one frame per image with exactly one barcode at the bottom (or maybe the top)?

The way this symbology is defined, I expect it will lead to a lot of false positives when presented with other types of barcodes. So it would be required to disable it in a standard setting. In that regard it might end up being treated as a second class citizen but if that is fine with you, I would definitely consider merging a PR provided that it does not affect the performance overall (which is a given if it is disabled by default).

**Comment by Merinorus** [comment]
Most of the time, the DSLR scanning system shows only the central part of the negative, to keep the film as flat as possible.. Like this:
![image](https://github.com/zxing-cpp/zxing-cpp/assets/3023499/3b6ea9cd-90cf-4cf6-a9fa-28a8f63a1c70)

So we can view these barcodes in a few systems, or just by taking a simple picture with our smartphone.

There are also two barcodes per image. If we take the negative on the first picture of my post, there are two barcodes with a difference in the "frame number" section. They are numbered every half-frame number (1, 1A, 2, 2A...). For example:

![Sans titre-1](https://github.com/zxing-cpp/zxing-cpp/assets/3023499/2db6330e-5ec5-43d0-9d81-d6b9908ef9db)
- Barcode 1:
  - Frame number (green): 0b010110 = 22
  - Flag (blue): 0b1 = second half of the frame -> **22A**
- Barcode 2:
  - Frame number (green): 0b010110 = 23
  - Flag (blue): 0b0 = first half of the frame -> **23**

They put a number every half-frame to avoid losing information with the notching machines, just in case one of the barcodes is cut.
Note: sometimes the half-frame number is aligned differently. It might be at the very beginning of the frame, in the middle, or even in arbitrary places (and all the other frames are shifted the same way)...

Anyway, you're right: these barcodes are on negative films only and their detection should be disabled by default. I agree with that!

**Comment by axxel** [comment]
So there is no open question left for me to answer, right?

**Comment by Merinorus** [comment]
No... I have to review the other barcodes to check which one would have a logic similar to this one, so I can take examples in the code, but I have not much experience with C++... So it's a nice functional AND technical challenge for me!
I'll keep you informed if someday I can sort it out.

**Comment by axxel** [comment]
Very good. Just to give you a first pointer where to start looking: I recommend the `ODITFReader.cpp` file. Especially the `FindLeftGard()` -> `IsRightGuard()` structure should be applicable here. The middle part is different to probably all the existing symbologies because they are based on the idea that you have a fixed number of bars+spaces per 'character' or smallest information unit. In your case, you have just a binary number directly encoded into bars and spaces. So here you might find that the model of the `PatternView` is not really ideal for reading of your bits.

Good luck :).

**Comment by axxel** [comment]
You can't compare this stacked 1D code with a 2D code like QRCode or Aztec because those have specific finder patterns that can be used locate the symbols or at least a potential symbol. Your situation is more comparable to the stacked DataBar symbologies. You can have a look at those and see how they use the `std::unique_ptr<DecodingState>&` parameter to store state from one line to the next. What you could do with that is the following:
 * first only look for the clock part which is a fixed pattern and this pattern is so long that it is actually relatively safe in terms of false positives
 * for each clock pattern you find, store the beginning and end in your `DecodingState`
 * if you find a startpattern, check if it starts roughly at the same x-position as a clock-pattern and only then treat it as such

This only works if you come across the clock pattern before the start pattern, which is the case since, the image is scanned from the center line towards the borders. If you are certain that any such barcode can only show up near the bottom of the image that could also be used so speed up the process and limit false positives. Thinking about that, it might even make sense to actually model it after a 2D symbology after all and pay the extra costs of sampling those lines you are interested in again (see `BinaryBitmap::getPatternRow`). But this is only useful if this assumption (look only at the bottom) holds true in all cases you are interested in.

**Comment by Merinorus** [comment]
Thank you for the hint. Indeed, finding the clock before helps a lot in avoiding false positives!
It seems almost good to me. I have to crop the image and it detects one code per row, but I think it would be fine for the application.
However, since this type of code is very flattened, the image needs to be almost perfectly horizontally aligned.
Here is a demo with 1D only, with your method of storing the decoding state:

https://github.com/zxing-cpp/zxing-cpp/assets/3023499/4ef04dba-a242-4bb9-adb9-434f32341014

I'm thinking about detecting the 2D pattern so I can:
- detect if the pattern is rotated
- apply the needed rotation angle to the image, to increase the number of detected rows.

Is it reasonable or should I keep it as it is?

**Comment by Merinorus** [comment]
Hello!
I have made a first attempt, some parts may need to be reworked, let me know what you think...
Thanks,

**Comment by axxel** [comment]
Do you think with the merging and subsequent tuning of your PR this issue can be closed now?

**Comment by Merinorus** [comment]
I just looked at your commits. Honestly, there are some implementation details I couldn't understand quickly but it looks good to me. I tried the sample Android app again and it works perfectly.

I reread the Kodak patent: nothing about the output syntax. I think the `XXX-Y/ZZ[A]` format should be okay: it's easy enough to parse both ways (if someday we want to implement the writer).

Overall I'm really happy with the result, thank you for your time!
